compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4544,6 +4544,28 @@ private void setResultTypeForWaitForAllExpr(BLangWaitForAllExpr waitForAllExpr,
             case TypeTags.TYPEREFDESC:
                 setResultTypeForWaitForAllExpr(waitForAllExpr, Types.getReferredType(expType), data);
                 break;
+            case TypeTags.UNION:
+                LinkedHashSet<BType> unionMemberTypes = ((BUnionType) expType).getMemberTypes();
+                int prevErrorCount = this.dlog.errorCount();
+                this.dlog.mute();
+                for (BType memberType : unionMemberTypes) {
+                    data.expType = memberType;
+                    this.dlog.resetErrorCount();
+                    setResultTypeForWaitForAllExpr(waitForAllExpr, memberType, data);
+                    if (this.dlog.errorCount() == 0) {
+                        this.dlog.setErrorCount(prevErrorCount);
+                        this.dlog.unmute();
+                        data.expType = expType;
+                        return;
+                    }
+                }
```

Comment:
This doesn't handle ambiguities, does it?

E.g.,

```ballerina
type Result record {
    string|error a;
    string|error b;
};

type Result2 record {
    any|error a;
    any|error b;
};

public function main() {
    
    worker WA returns string|error {
        return "";
    }

    worker WB returns string|error {
        return "";
    }

    Result2|Result x = wait {a: WA, b: WB};
}
```

Static type used for the created record value depends on what is matched first, which is not how we construct values. If there's more than one type that can be used, we log an error.

Reply:
Did we look at what the spec says for wait with unions?

https://ballerina.io/spec/lang/master/#multiple-wait-action says

> The static type of the multiple-wait-action is a record where the static type of each field is the eventual type of the corresponding wait-future-expr.

It doesn't seem to talk about using the contextually-expected type. So we probably should be creating a new record type for the wait-action, even for the non-union case, and then just type check it against the expected type. This will fix the original issue too, since the record type we create will be a subtype of the expected union type.

We can maybe raise this in the spec and get it clarified, because I believe the following will no longer work.

```ballerina
type Result record {
    any|error x;
};

public function main() {
    worker W returns string|error {
        return "hello";
    }

    Result res = wait {x: W};

    // Doesn't panic at runtime in the current impl., since we use `Result` as the type of the constructed mapping value.
    // But if the type is `record {| string|error x; |}`, this will panic.
    res.x = 1;
}
```